lnwallet/chancloser/rbf_coop_transitions.go

```diff
@@ -0,0 +1,1109 @@
+			NextState: &ShutdownPending{
+				prevState:    c,
+				inputEvents:  shutdownTransition,
+				IdealFeeRate: fn.Some(msg.IdealFeeRate),
+				ShutdownScripts: ShutdownScripts{
+					LocalDeliveryScript: shutdownScript,
+				},
+			},
+			// TODO(roasbeef): type alias
+			NewEvents: fn.Some(protofsm.EmittedEvent[ProtocolEvent]{
+				ExternalEvents: fn.Some(daemonEvents),
+			}),
+		}, nil
+
+	// When we receive a shutdown from the remote party, we'll validate the
+	// shutdown message, then transition to the ChannelFlushing state.
+	// We'll also emit similar events like the above to send out shutdown,
+	// and also disable the channel.
+	case *ShutdownReceived:
+		chancloserLog.Infof("ChannelPoint(%v): received shutdown msg")
+
+		// Validate that they can send the message now, and also that
+		// they haven't violated their commitment to a prior upfront
+		// shutdown addr.
+		err := validateShutdown(
+			env.ThawHeight, env.RemoteUpfrontShutdown, msg,
+			env.ChanPoint, env.ChainParams,
+		)
+		if err != nil {
+			chancloserLog.Errorf("ChannelPoint(%v): rejecting "+
+				"shutdown attempt: %v", err)
+
+			// TODO(roasbeef): emit disconnect event?
+			return nil, err
+		}
+
+		// If we have an upfront shutdown addr we'll use that,
+		// otherwise, we'll generate a new delivery script.
+		shutdownAddr, err := env.LocalUpfrontShutdown.UnwrapOrFuncErr(
+			env.NewDeliveryScript,
+		)
+		if err != nil {
+			return nil, err
+		}
+
+		chancloserLog.Infof("ChannelPoint(%v): sending shutdown msg "+
+			"at next clean commit state", env.ChanPoint)
+
+		// Now that we know the shutdown message is valid, we'll obtain
+		// the set of daemon events we need to emit. We'll also specify
+		// that once the message has actually been sent, that we
+		// generate receive an input event of a ShutdownComplete.
+		daemonEvents, err := sendShutdownEvents(
+			env.ChanID, env.ChanPoint, shutdownAddr,
+			env.ChanPeer,
+			fn.Some[ProtocolEvent](&ShutdownComplete{}),
+			env.ChanObserver,
+		)
+		if err != nil {
+			return nil, err
+		}
+
+		chancloserLog.Infof("ChannelPoint(%v): disabling incoming adds",
+			env.ChanPoint)
+
+		// We just received a shutdown, so we'll disable the adds in
+		// the outgoing direction.
+		if err := env.ChanObserver.DisableIncomingAdds(); err != nil {
+			return nil, fmt.Errorf("unable to disable incoming "+
+				"adds: %w", err)
+		}
+
+		// We'll also record that we arrived at the ChannelFlushing
+		// state via a ShutdownReceived event, which means this was a
+		// locally initiated shutdown.
+		shutdownTransition := fn.NewRight[SendShutdown](*msg)
+
+		return &CloseStateTransition{
+			NextState: &ShutdownPending{
+				prevState:   c,
+				inputEvents: shutdownTransition,
+				ShutdownScripts: ShutdownScripts{
+					LocalDeliveryScript:  shutdownAddr,
+					RemoteDeliveryScript: msg.ShutdownScript,
+				},
+			},
+			NewEvents: fn.Some(protofsm.EmittedEvent[ProtocolEvent]{
+				ExternalEvents: fn.Some(daemonEvents),
+			}),
```

Comment:
So when you emit something, it can be an internal and/or external event. Emitting events in general is optional. 

You suggest to eliminate the outer layer, and keep if for the attributes?

Reply:
That would be my instinct, yes. You can recover the outer None functionality with two inner Nones. It would eliminate the "useless case" of Some(None, None) which is an example of where the data model doesn't reflect the contours of the actual problem.